### ISSUE TITLE
dm-lab in Keras

### ISSUE THREAD

**Issue by uahic** [deepmind/lab#177]
Heya, 

I'm currently re-implementing the vector-based navigation architecture of deep-mind's nature paper for tensorflow 2. However, it seems there is an issue regarding the environment of the python API when used with the Keras model API. In particular, I do something like:

```
...
observations = ['RGB', 'DEBUG.POS.ROT', 'DEBUG.POS.TRANS']
env = deepmind_lab.Lab(level, observations,
                           config=lab_config, renderer='hardware')
env.reset()
...
    def generate_batch(env, batch_size=10):
        while True:
            ...
            obs = env.observations()
            ...
            x, y = ... some computation 
            env.reset()
            ...
            yield x, y


model.fit_generator(
generate_batch(env, batch_size=10),
steps_per_epoch=1000,
verbose=1,
use_multiprocessing=False
)

```

depending whether i use env.reset() or not it sometimes drops error (related to: https://github.com/deepmind/lab/issues/134). Even if it does not explicitely complains, it stops always within the first batch. Since Tensorflow 2 seems to convert the generate_batch function into some eager-execution/autograph stuff, I am quite confused whether the env object gets either copied or something else nasty going on under the hood. I hope someone has more clue than I do :) 

The agent I am using is performing random actions (so it should not be the same state-sequence all the time, especially because the rnd seed is not statically set)

**Comment by tkoeppe** [comment]
What does "drops error" mean?

Can you try again with the software renderer? If it's related to #134, that should be a way to narrow it down.

**Comment by uahic** [comment]
@tkoeppe I'm getting the following error: 
```
EGL ERROR: 0x3002 file:engine/code/deepmind/headless_egl_glimp.c, line:67
```
when env.reset() is called right after
```
    observations = ['RGB', 'DEBUG.POS.ROT', 'DEBUG.POS.TRANS']
    env = deepmind_lab.Lab(level, observations,
                        config=lab_config, renderer='software')
    env.reset()
```

if env.reset() is done in the generate_batch() function OR using software-rendering, it doesnt seem to happen. My current guess is that using this generator function as a closure - which is passed to keras and probably converted - is problematic for the env reference.  I lack knowledge of the internals of dmLab and Keras to really judge at this point

**Comment by tkoeppe** [comment]
The EGL error looks like what might happen if you're trying to create multiple environments in the same process; I *think* EGL does not support that. I'm not actually 100% sure here, but my vague recollection is that only the software renderer can be instantiated multiple times within one process. (Maybe someone could confirm this in an isolated, toy setup?)

Are you calling `deepmind_lab.Lab(...)` in an inner loop somehow? The way this constructor works is by making file [copies of the main DSO](https://github.com/deepmind/lab/blob/master/public/dmlab_so_loader.cc#L144) and using `dlopen` on these iterated copies; it's possible that if you're in some unexpected situation that that doesn't work. Maybe try ascertaining whether those file operations succeed?

**Comment by uahic** [comment]
The structure of this script is rather simple, here is a rough outline:

```
def train_viz_module(...):
    env = deepmind_lab.Lab(level, observations,
                        config=lab_config, renderer='software')
    env.reset()

    def generate_batch(...):
        while True:
            // do stuff
            // get observations
            // process observations
            x_in, y_out = ...
            yield x_in, y_out

    model = ... some keras custom model            
    model.compile(...)
    model.fit_generator(generate_batch(...), ... other args)
if __name__ == '__main__':
    train_viz_module(...)
```

**Comment by tkoeppe** [comment]
I can't help with the specific details, but I can give you some debugging directions.

The Python `Reset` function's [implementation](https://github.com/deepmind/lab/blob/master/python/dmlab_module.c#L311) ultimately calls the EnvCApi's [start](https://github.com/deepmind/lab/blob/master/python/dmlab_module.c#L338) function, which is implemented [here](https://github.com/deepmind/lab/blob/master/engine/code/deepmind/dmlab_connect.c#L775).

When you say "it doesnt seem to happen", you can try and see if any of those calls actually happen (e.g. by setting a breakpoint, or even just by adding some `printf`s). That should let get a more focused description of your problem.

**Comment by tkoeppe** [comment]
@uahic: Any updates on this?

**Comment by uahic** [comment]
@tkoeppe Nope but I switched to replay buffers for reinforcement learning training and sample from them. I didnt thought about this possibility in the first place, sorry!

**Comment by tkoeppe** [comment]
Ah, OK, no worries! Can we close this issue then, or is there anything else you'd like to investigate?

**Comment by uahic** [comment]
We can close this issue for now, thanks again for your help! If anything critical pops up I will let you know
